---
# Seminar 1 - Grundlagen
---

### Hinweis: Speichern eines Notebooks immer über: STRG+S/ CMD+S oder das Disketten-Symbol. Die Option File --> Save funktioniert bei Notebooks nicht!!


## 1. Podcast-Sendezeit

Sie haben es geschafft und konnten sich einen der begehrten Praktikumsplätze im Team des Podcast-Newcomers "OR" ergattern. Ihre erste Aufgabe besteht darin, herauszufinden, wie viele Sendeminuten der Podcast wöchentlich hat.

1. Eine Podcastfolge dauert im Durchschnitt 70 Minuten. Dies soll aber variabel anpassbar sein. (Variable)
2. Zusätzlich werden pro Folge 3 Werbeblöcke à 3 Minuten eingespielt.
3. Wöchentlich werden 2 Podcasts veröffentlicht.
4. Die Redaktionschefin bittet Sie, das Ergebnis in Stunden gerundet auf zwei Nachkommastellen auszugeben.
5. Des Weiteren soll nicht nur einfach eine Zahl ausgegeben werden, sondern ein Satz: "Bei einer Podcastlänge von XX Minuten ergeben sich Y.YY Stunden Sendezeit pro Woche."

### a.)

Schreiben Sie bitte ein Skript, welches die Berechnung übernimmt und geben Sie das Ergebnis aus.


In [2]:
podcast_length = 70
advertisement = 3
podcast_perweek = 2

podcast_total = float(podcast_perweek*(podcast_length + 3*advertisement))
print(f'Bei einer Podcastlänge von {podcast_length} Minuten, drei Werbeblöcken á {advertisement} Minuten'+
      f'und {podcast_perweek} Podcasts pro Woche, ergeben sich {round(podcast_total/60,2)} Stunden Sendezeit pro Woche.')

Bei einer Podcastlänge von 70 Minuten, drei Werbeblöcken á 3 Minutenund 2 Podcasts pro Woche, ergeben sich 2.63 Stunden Sendezeit pro Woche.


### b.)

Informieren Sie sich bitte im Internet, welche weiteren Möglichkeiten es gibt, Zeichenketten für eine Ausgabe zu formatieren.

**Tipp:**
<details>

Schreiben Sie den Ausgabebefehl von Aufgabe a) als "".format() und f""-Ausgabe.
</details>

In [3]:
print('Bei einer Podcastlänge von {} Minuten, drei Werbeblöcken á {} Minuten und {} Podcasts pro Woche, ergeben sich {} Stunden Sendezeit pro Woche.'.format(podcast_length, advertisement, podcast_perweek, round(podcast_total/60,2)))

Bei einer Podcastlänge von 70 Minuten, drei Werbeblöcken á 3 Minuten und 2 Podcasts pro Woche, ergeben sich 2.63 Stunden Sendezeit pro Woche.


Erwarteter Output:

    Bei einer Podcastlänge von 70 Minuten ergeben sich ...

## 2. Chaos in der Mailingliste

Die Hörer des Podcasts haben die Möglichkeit, sich für einen Newsletter anzumelden. So erfahren sie schnellstmöglich, wann es eine neue Podcast-Folge gibt. Dabei ist aufgefallen, dass manche Mailadressen den Namen erst hinter dem @-Zeichen haben. 

### a.)

Entwickeln Sie bitte einen Python-Ausdruck, der die Namen hinter dem @-Zeichen als Elemente einer Liste ausgibt. 

In [4]:
mail = "service@klara_Klarna.com"

test = mail.split('@')[1].split('.')[0].split('_')
print(test)

['klara', 'Klarna']


Erwarteter Output:

    ['klara', 'Klarna']

### b.)

Nachdem Sie nun die Namen der AbonnetInnen zu den jeweiligen E-Mail-Adressen zuordnen können, stellt die Chefredakteurin Ihnen eine neue Aufgabe. Überführen Sie bitte die folgende Tabelle in ein Dictionary bestehend aus den Namen als Keys und den E-Mail-Adressen als Values:

| Name      	|   	| E-Mail                          	|
|:-----------	|---	|:---------------------------------	|
| Klara     	|    	| service@klara_Klarna.com        	|
| Sebastian 	|    	| Sebastian_Duesentrieb@gmail.com 	|
| Britta    	|    	| bwiebritta@yahoo.com            	|
| Klaus     	|    	| Klaus@peter.de                  	|

Legen Sie dazu zunächst zwei Listen an, die Sie dann zusammenführen können.

**Tipp:**
<details>

Nutzen Sie die Funktion `enumerate()` um die Indizes und Elemente einer Liste zu verwenden.
</details>

In [5]:
names = ['Klara', 'Sebastian', 'Britta', 'Klaus']
mails = ['service@klara_Klarna.com', 'Sebastian_Duesentrieb@gmail.com', 'bwiebritte@yahoo.com', 'Klaus@peter.com']

kundendaten = dict()
for i, name in enumerate(names):
    kundendaten[name] = mails[i]
print(kundendaten)

{'Klara': 'service@klara_Klarna.com', 'Sebastian': 'Sebastian_Duesentrieb@gmail.com', 'Britta': 'bwiebritte@yahoo.com', 'Klaus': 'Klaus@peter.com'}


## 3. Anpassung der Werbeeinnahmen

Die Freude im Podcast-Team ist rießig. Zuletzt kamen immer mehr interessierte ZuhörerInnen hinzu und so konnte die Zuhörerzahl in den letzten 3 Monaten um 400% gesteigert werden. Der kaufmännische Leiter des Teams will daher die Einnahmen durch die angebotenen Werbepakete "optimieren". 

Im Moment existieren vier Werbepakete für den Podcast. Bisher wurden immer kundenspezifische Preise bis zu einer Höchstgrenze für jedes Paket verhandelt.

1. Paket "Einleitung", Kosten bis 350 €
2. Paket "Mitte", Kosten bis 300 €
3. Paket "Komplett", Kosten bis 600 €
4. Paket "Zwischendurch", bis 800 €

### a.)

Der kaufmännische Leiter hat beschlossen, dass aufgrund der deutlich vergrößerten Reichweite folgende Preisanpassungen wohl gerechtfertigt sind:
Die Pakete bis 400 € werden um 10% angehoben, die Pakete zwischen 400 und 600 € um 20% und die Pakete über 600 € um 25%.

Bitte setzen Sie die Preisanpassung in einem Pythonskript um, sodass der neue Preis in Abhängigkeit eines variablen Eingabepreises ausgegeben wird.

In [6]:
input_price = 333

if input_price <= 400:
    new_price = input_price*1.1
elif 400 < input_price <= 600:
    new_price = input_price*1.2
else:
    new_price = input_price*1.25

print(f'Der angehobene Preis beträgt {round(new_price,2)} Euro.')


Der angehobene Preis beträgt 366.3 Euro.


Erwarteter Output:

    366.3

### b.)

Durch den Erfolg des Podcasts sind nun auch die Zeiten der kundenspezifischen Preise und Verhandlungen vorbei. Jedes Paket hat ab jetzt den oben genannten Maximalpreis als Festpreis.

Bitte speichern Sie die Paketpreise in eine Liste und generieren Sie mithilfe dieser Liste eine neue Liste, in der die neuen Preis zu finden sind.

In [7]:
oldPrice = [350, 300, 600, 800]
newPrice = []

for price in oldPrice:
    if price <= 400:
        newPrice.append(int(price*1.1))
    elif price <= 600:
        newPrice.append(int(price*1.2))
    else:
        newPrice.append(int(price*1.25))

newPrice

[385, 330, 720, 1000]

Erwarteter Output:

    [385, 330, 720, 1000]

### c.) 

Nun ist Kassensturz angesagt! Ihr kaufmännischer Leiter will jetzt wissen, was die Preisanpassung in den vergangenen Wochen gebracht hat. 

Dafür soll eine Funktion geschrieben werden, welcher zwei Listen (Preise und Verkaufszahlen) als Parameter übergeben werden und diese Funktion, dann den **Umsatz** zurückgibt.


Hinweis: Sie können annehmen, dass die Indizierung der beiden Listen identisch ist, d.h. der erste Eintrag in der Preisliste gehört zum ersten Eintrag in der Liste der Verkaufszahlen.

In [8]:
sales = [3, 5, 7, 4]
oldPrice = [350, 300, 600, 800]
newPrice = [385, 330, 720, 1000]

def calc_revenue(prices, sales):
    revenue = 0
    for i in range(len(sales)):
        revenue += sales[i]*prices[i]
    return revenue

print(f'Umsatz alte Preise: {calc_revenue(oldPrice, sales)} Euro')
print(f'Umsatz neue Preise: {calc_revenue(newPrice, sales)} Euro')
print('--------------------------------')
print(f'Umsatzsteigerung: {calc_revenue(newPrice, sales) - calc_revenue(oldPrice, sales)} Euro')

Umsatz alte Preise: 9950 Euro
Umsatz neue Preise: 11845 Euro
--------------------------------
Umsatzsteigerung: 1895 Euro


Erwarteter Output:

    1895

### d.)

Informieren Sie sich bitte über die Möglichkeiten der lambda-Funktion. Setzen Sie bitte anschließend die Funktion **Umsatz** als lambda-Funktion um.

In [9]:
calc_revenue_lambda = lambda prices, sales: sum([sales[i]*prices[i] for i in range(len(sales))])
print(f'Umsatz alte Preise (Lambda): {calc_revenue_lambda(oldPrice, sales)} Euro')
print(f'Umsatz neue Preise (Lambda): {calc_revenue_lambda(newPrice, sales)} Euro')
print('--------------------------------')
print(f'Umsatzsteigerung (Lambda): {calc_revenue_lambda(newPrice, sales) - calc_revenue_lambda(oldPrice, sales)} Euro')

Umsatz alte Preise (Lambda): 9950 Euro
Umsatz neue Preise (Lambda): 11845 Euro
--------------------------------
Umsatzsteigerung (Lambda): 1895 Euro


Wie können lambda-Funktionen noch eingesetzt werden?

## 4. Expansionswahnsinn

Vom neuerlichen Erfolg des Podcasts getrieben schlägt der kaufmännische Leiter nun vor, mit einem englischsprachigen Ableger des Podcasts in die USA zu expandieren. Die Chefredakteurin ist noch skeptisch und beauftragt Sie mit der Analyse des Marktpotenzials. In der Zielgruppe des Podcasts sind laut Studien die meisten Personen zwischen 25 und 35 Jahren alt.

### a.)

Finden Sie daher heraus, wie viele US-BürgerInnen derzeit in der relevanten Altergruppe sind (einschließlich der genannten Grenzen). Leider hat die Chefredakteurin nur eine Tabelle auftreiben können, die alle Geburten nach Bundesstaaten und Vornamen gliedert. Versuchen Sie die benötigten Informationen aus diesem Datensatz zu generieren(births_red.csv)!

In [10]:
import pandas as pd

us_people = pd.read_csv('../data/births_red.csv', sep=',')
us_people.head()

,state,Gender,Year,Count
0,AK,F,1961,2693
1,AK,F,1962,2693
2,AK,F,1963,2706
3,AK,F,1964,2520
4,AK,F,1965,2286


In [11]:
print(us_people.shape)
print(us_people.columns)
print(us_people.dtypes)

(6018, 4)
Index(['state', 'Gender', 'Year', 'Count'], dtype='object')
state     object
Gender    object
Year       int64
Count      int64
dtype: object


In [12]:
us_people.Count[(2025-us_people['Year']>=25) & (2025-us_people.Year<=35)].sum()

np.int64(36620130)

In [13]:
count = 0
for year in us_people['Year']:
    if 25 <= 2025-year <= 35:
        count += us_people['Count'][us_people['Year']==year].sum()

print(count)

3735253260


Erwarteter Output:

    36620130

#### b.)

Die Chef-Redakteurin ist von Ihrem Ergebnis angetan. Um das Programm passgenau auf die meisten potenziellen ZuhörerInnen abzustimmen, bittet sie Sie um eine weitere Analyse. 

Bitte finden Sie heraus, welches Alter die meisten Personen innerhalb der relevanten Altersgruppe haben. Nutzen Sie dabei ein Dictionary, welches als Schlüssel das Alter und als Wert die jeweilige Personenanzahl beinhaltet.

In [18]:
people_per_age = {}    # dict()
for year in us_people['Year']:
    if 25 <= 2025-year <= 35:
        people_per_age[2025-year] = us_people.Count[us_people['Year']==year].sum()

print(people_per_age)

age_max_people, max_people = max(people_per_age.items(), key = lambda x: x[1])  # lambda x:x[1] determines that max is found in the second element of the tuple (key, value)
print(f"Das häufigste Alter ist {age_max_people} mit {max_people} Personen.")

{35: np.int64(3567577), 34: np.int64(3503407), 33: np.int64(3442904), 32: np.int64(3364630), 31: np.int64(3307859), 30: np.int64(3249949), 29: np.int64(3225793), 28: np.int64(3194233), 27: np.int64(3230662), 26: np.int64(3234596), 25: np.int64(3298520)}
Das häufigste Alter ist 35 mit 3567577 Personen.


Erwarteter Output:

    35
    3567577

In [17]:
testing = people_per_age.items()
print(testing)
print(type(testing))

dict_items([(35, np.int64(3567577)), (34, np.int64(3503407)), (33, np.int64(3442904)), (32, np.int64(3364630)), (31, np.int64(3307859)), (30, np.int64(3249949)), (29, np.int64(3225793)), (28, np.int64(3194233)), (27, np.int64(3230662)), (26, np.int64(3234596)), (25, np.int64(3298520))])
<class 'dict_items'>
